vllm/worker/model_runner.py

```diff
@@ -1576,6 +1577,24 @@ def execute_model(
         else:
             model_executable = self.model
 
+        # Receive KV cache in distributed KV cache transfer setting
+        # In disagg prefill setting, it will also recv hidden states and bypass
+        # model forwarding
+        # In KV cache database setting, it will change the model input so that
+        # we can skip prefilling on tokens that successfully received KV caches
+        # NOTE: The receive operation is blocking
+        bypass_model_exec = False
+        if self.need_recv_kv(model_input, kv_caches):
+            hidden_or_intermediate_states, bypass_model_exec, model_input = \
+                get_disagg_group().recv_kv_caches_and_hidden_states(
```

Comment:
Just curious. Is this recv function blocking? 

It seems to me that adding an independent thread that constantly waiting for kvcache might hide the communication cost, but I am not sure if this is feasible under current block manager logic.

Reply:
Yeah currently it is blocking (and it is an intentional design choice). 
Reasons behind this design:
- Transfer delay is not significant (~5ms). Most of the overheads in the implementation is tokenizing the whole request twice and de-tokenizing the first token twice (~20ms). Optimizing other overheads will give us perf improvement.
- Non-blocking KV transfer requires adding GPU buffer at decoding vLLM side. Since decoding is memory-bounded, this is not ideal (as buffering reduces # of available GPU memory for decoding KV caches).

Reply:
> * Transfer delay is not significant (~5ms). Most of the overheads in the implementation is tokenizing the whole request twice and de-tokenizing the first token twice (~20ms). 

Thanks for this helpful information! I agree with your observation. Skipping sampling & output processor in prefill side and sending kvcache immediately after hidden states will be more significant.